### DISCUSSION TITLE
How to deal with non QWERTY layouts?

### DISCUSSION THREAD

**Discussion by MarkuBu** [helix-editor/helix#8477] [Q&A]
Hi, I want to use Helix more often, but I'm struggling with the keyboard shortcuts. 

On my laptop I use a normal QWERTZ (German) keyboard, but on my main machine I have a Redox keyboard with a special layout. hjkl are distributed everywhere, so it is impossible to use the muscle memory from the standard keyboard on the Redox.

What should I do? Remap everything to fiit the Redox keyboard or learn two different mappings?

**Comment by the-mikedavis** [comment]
Most of the maintainers don't use Qwerty actually: I use Dvorak, @pascalkuthe uses Colemak, @archseer uses Norman. With the exception of a few historical keys like `hjkl` most keys are chosen for their ease of memory rather than position in Qwerty. For example `y` to yank, `p` to paste, `u` to undo, etc.. So unless any bindings are really tricky to hit in your layout I would recommend using the default keymap as-is.

**Reply by MarkuBu** [reply]
I think I have two options: use the cursor keys as usual or I can add a qwerty layout to my keyboard and switch between qwerty (normal mode) and my layout (insert mode)

**Reply by kirawi** [reply]
You can rebind the cursor keys; some of the maintainers also do that IIRC.

**Reply by pascalkuthe** [reply]
I am actually German myself. Safe your hands and just use qwerty permanently. QWERTY isn't gperfect but it's a huge improvement over QWERZ. The German layout is unsuitable for programming with all the altgr combos

**Reply by MarkuBu** [reply]
> I am actually German myself. Safe your hands and just use qwerty permanently. QWERTY isn't gperfect but it's a huge improvement over QWERZ. The German layout is unsuitable for programming with all the altgr combos

I use the Redox keyboard with the PUQ layout (See adnw.de). I have the symbols on a separate layer. Also cursor keys.

And I agree: qwertz is horrible for programming.

**Reply by cd-a** [reply]
Having 2 layouts on different machines is very confusing, it will be quite a challenge to achieve perfect muscle memory from it.

I have a Glove80 keyboard which is set up with Colemak, but whenever I use my laptop without my external keyboard, I just switch the layout to Colemak as well, and it works just great.

I recommend you pick one layout and use it across all your machines. Just don't look at the keys, since it will be all mis-labeled then. Oh, and as @pascalkuthe mentioned, QWERTZ is probably the worst possible one for programming, it's really a pain. 

Use the one that you're most comfortable with. And if you're worried about using a non-german layout and missing out on äöü, don't be. In my case, I don't even write German much. But when I do, there's always the lazy way to write as ae oe ue, but even better - you can compose characters very easily. Alt-u + u = ü, and so on. It's really fast after a bit of practice, so don't sweat it.

So I recommend optimise the layout to your most important task, which I assume is coding.

**Reply by MarkuBu** [reply]
> I recommend you pick one layout and use it across all your machines. Just don't look at the keys, since it will be all mis-labeled then. 

That's not possible because my main keyboard is a Redox keyboard and the layout is optimized for it. I heavily use the thumb clusters for modifiers. And even the layout itself doesn't really work well on regular keyboards.

**Comment by johannesrave** [comment]
edit: i'm sorry, i have found the answer to my uncertainty below in... the official documentation under [remapping](https://docs.helix-editor.com/remapping.html), with the example given as `a = "move_char_left" # Maps the 'a' key to the move_char_left command`. i'd still like to know about other keys used because of their position.

original:
going back to the original question - i'm trying to become proficient with the bone/neo2 layout, and i face a similiar challenge to OP. 

it sounds like the keymap can be categorized into keys that are chosen for mnemonic considerations, such as the mentioned *y*ank, *i*nsert, *o*pen, *g*oto and others - i'm completely fine with just leaving those be with whatever layout. 

but then there are keys the only make sense because of their position, such as the cursor-keys on the home row. i'd really like to remap those, and am unsure how.
i also wonder whether other keys used in helix also follow that positional reasoning and fall into that category - do any jump to mind?

as a backup, this layout has very functional cursor-keys mapped to layer 4, if it's too much of a hassle, i might use those.

**Reply by MarkuBu** [reply]
If you are familiar with bone/Neo2 you should be also familiar with their layers. 

I use the navigation layer heavily instead of hjkl.

But you should skip Neo2. It is not so godd as expected. Look at adnw.de for an alternative like Koy for regular keyboards. And if you want to go crazy get a Redox or Moonlander. But then you should look for PUQ.
